internal/info/info.go

```diff
@@ -54,43 +69,84 @@ type StackTrace struct {
 }
 
 var (
-	Version      = "v0.0.1"
-	GitCommit    = "master"
+	// injected from build script
+	Version           = "v0.0.1"
+	GitCommit         = "master"
+	BuildTime         = ""
+	GoVersion         string
+	GoOS              string
+	GoArch            string
+	CGOEnabled        string
+	NGTVersion        string
+	BuildCPUInfoFlags string
+
 	Organization = "vdaas"
 	Repository   = "vald"
-	BuildTime    = ""
 
-	GoVersion  string
-	GoOS       string
-	GoArch     string
-	CGOEnabled string
+	reps = strings.NewReplacer("_", " ", ",omitempty", "")
 
-	NGTVersion string
+	once         sync.Once
+	infoProvider InformationProvider
+)
 
-	BuildCPUInfoFlags string
+// Init initializes Detail object only once.
+func Init(name string) {
+	once.Do(func() {
+		infoProvider, _ = New(WithServerName(name))
+	})
+}
 
-	reps = strings.NewReplacer("_", " ", ",omitempty", "")
+// New initialize and return the information provider or any error occurred.
+func New(opts ...Option) (InformationProvider, error) {
+	i := &info{
+		detail: Detail{
+			Version:           Version,
+			GitCommit:         GitCommit,
+			BuildTime:         BuildTime,
+			GoVersion:         GoVersion,
+			GoOS:              GoOS,
+			GoArch:            GoArch,
+			CGOEnabled:        CGOEnabled,
+			NGTVersion:        NGTVersion,
+			BuildCPUInfoFlags: strings.Split(strings.TrimSpace(BuildCPUInfoFlags), " "),
+		},
+	}
 
-	once sync.Once
+	for _, opt := range append(defaultOpts, opts...) {
+		if err := opt(i); err != nil {
+			werr := errors.ErrOptionFailed(err, reflect.ValueOf(opt))
 
-	detail Detail
-)
+			e := new(errors.ErrCriticalOption)
+			if errors.As(err, &e) {
+				log.Error(werr)
+				return nil, werr
+			}
+			log.Warn(werr)
+		}
+	}
```

Comment:
should we validate `rtCaller ` or `rtFuncForPC ` ?

Reply:
what do you think? I put it in `defaultOptions`  and also we have nil validation when setting the field so it should never be nil...

Reply:
I got it, thank you.